kbase-extension/static/kbase/js/util/jobLogViewer.js

```diff
@@ -625,146 +666,289 @@ define([
-            kblogLine.appendChild(wrapperDiv);
-
-            return kblogLine;
+            const errorSuffix = line.isError ? '--error' : '';
+            return div(
+                {
+                    class: `${cssBaseClass}__line_container${errorSuffix}`,
+                },
+                div(
+                    {
+                        class: `${cssBaseClass}__flex_wrapper`,
+                    },
+                    [
+                        div(
+                            {
+                                class: `${cssBaseClass}__line_number${errorSuffix}`,
+                            },
+                            line.lineNumber
+                        ),
+                        div(
+                            {
+                                class: `${cssBaseClass}__line_text${errorSuffix}`,
+                            },
+                            line.text
+                        ),
+                    ]
+                )
+            );
         }
 
         function getLogPanel() {
             return ui.getElement('log-panel');
         }
 
         /**
-         * onUpdate callback function (under model)
+         * Render the log
+         *
+         * This replaces the log wholesale with the content of model.getItem('lines')
+         * rather than keeping track of which lines have and haven't been rendered.
          */
-        function render() {
+        function renderLog() {
             const lines = model.getItem('lines');
+            if (!lines || lines.length === 0) {
+                ui.setContent('log-panel', 'No log entries to show.');
+                return;
+            }
 
-            if (lines && lines.length > 0) {
-                const panel = getLogPanel();
-                panel.innerHTML = '';
-                lines.forEach((line) => panel.appendChild(buildLine(line)));
+            const panel = getLogPanel();
+            panel.innerHTML = lines.map((line) => buildLine(line)).join('\n');
 
-                // if we're autoscrolling, scroll to the bottom
-                if (fsm.getCurrentState().state.auto || scrollToEndOnNext) {
-                    panel.scrollTo(0, panel.lastChild.offsetTop);
-                    scrollToEndOnNext = false;
-                }
+            // if we're autoscrolling, scroll to the bottom
+            if (fsm.getCurrentState().state.auto || scrollToEndOnNext) {
+                panel.scrollTo(0, panel.lastChild.offsetTop);
+                scrollToEndOnNext = false;
+            }
+            return;
+        }
+
+        function createJobStatusLines(jobState) {
+            const lines = Jobs.createJobStatusLines(jobState, showHistory);
+
+            return lines.map((line) =>
+                div(
+                    {
+                        class: `${cssBaseClass}__job_status_detail`,
+                    },
+                    line
+                )
+            );
+        }
+
+        function renderJobState(jobState) {
```

Comment:
This function displays the current job state, including history (i.e. queue time, run time) and any errors, if applicable. In developer mode, the current job state is also shown, as is the state of the widget.